src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -73,101 +74,147 @@
 #define STOP(str) stop(str);
 #define BIND(label) bind(label); BLOCK_COMMENT(#label ":")
 
+#ifdef ASSERT
+extern "C" void disnm(intptr_t p);
+#endif
+
 // Patch any kind of instruction; there may be several instructions.
 // Return the total length (in bytes) of the instructions.
-int MacroAssembler::pd_patch_instruction_size(address branch, address target) {
+int MacroAssembler::pd_patch_instruction_size(address insn_addr, address target) {
   int instructions = 1;
```

Comment:
I don't like the fact that you have to replicate this switch and associated validation logic in the two separate methods `pd_patch_instruction_size` and `target_addr_for_insn`. How about making both of them call a common auxiliary method which will either retrieve a target address or patch it? i.e. have

`int MacroAssembler::pd_patch_instruction_size(address insn_addr, address target)`
and
`address MacroAssembler::target_addr_for_insn(address insn_addr, uint32_t insn)`

both call method

`int MacroAssembler::fetch_or_patch_target_addr(address insn_addr, uint32_t insn, address &target, boolean do_patch)`

That means you can have one copy of the dispatch logic with common verification in each end case and variant action in each case depending on the value of `do_patch`.

It also means the explanatory comment above only nees to exist in one place at the head of method `fetch_or_patch_target_addr`

Reply:
n.b. I forgot to explain but I hope it is obvious that `target` is passed as a reference so it can be an input when `do_patch == true` and an output when `do_patch == false`. The return value is the instruction count in either case.

Reply:
> I don't like the fact that you have to replicate this switch and associated validation logic in the two separate methods `pd_patch_instruction_size` and `target_addr_for_insn`. How about making both of them call a common auxiliary method which will either retrieve a target address or patch it? i.e. have

I thought about that, but it ends up being rather complex. A function that does two different things depending on a boolean parameter is something of a code smell. I'll have another look.

Reply:
> Another possibility would be to have the dispatch logic simply return a unique enum value for each of the various distinguished leading instruction sub-cases and then have the retrieve and patch methods do a second switch on the returned enum value.

... of course, rather than rely on a switch I guess a pucker OO solution would use a virtual dispatch ... :-)